src/cli/commands/test/mod.rs

```diff
@@ -112,22 +151,120 @@ fn list_cairo_files(root: &PathBuf) -> Result<Vec<PathBuf>, String> {
 	.map(|cmd_output: ListOutput| cmd_output.files)
 }
 
-fn compile_and_list_entrypoints(path_to_code: PathBuf) -> Option<(PathBuf, PathBuf, Vec<String>)> {
-	match compile(&path_to_code) {
-		Ok(path_to_compiled) => match list_test_entrypoints(&path_to_compiled) {
-			Ok(entrypoints) => Some((path_to_code, path_to_compiled, entrypoints)),
-			Err(e) => {
-				eprintln!(
-					"Failed to list test entrypoints for file {}: {}",
-					path_to_compiled.display(),
-					e
-				);
-				None
+fn read_json_file(path: &PathBuf) -> Result<CacheJson, String> {
+	// Open the file in read-only mode with buffer.
+	let file = File::open(path).map_err(|op| format!("file does not exists {}", op))?;
+	let reader = BufReader::new(file);
+	let data =
+		serde_json::from_reader(reader).map_err(|op| format!("file does not exists {}", op))?;
+	return Ok(data);
+}
+
+fn compile_and_list_entrypoints(
+	cache: Result<CompiledCacheFile, String>,
+) -> Option<(PathBuf, PathBuf, Vec<String>)> {
+	match cache {
+		Ok(cache) => match cache.status {
+			CacheStatus::Cached => {
+				println!("Using cached compiled file");
+				let compiled_path = cache.path.clone();
+				let entrypoints =
+					list_test_entrypoints(&cache.path).expect("Failed to list entrypoints");
+				return Some((cache.path, compiled_path, entrypoints));
+			},
+			CacheStatus::Uncached => {
+				let compiled_path = compile(&cache.path).expect("Failed to compile");
+				let entrypoints =
+					list_test_entrypoints(&compiled_path).expect("Failed to list entrypoints");
+				return Some((cache.path, compiled_path, entrypoints));
 			},
 		},
-		Err(e) => {
-			eprintln!("{}", e);
-			None
+		Err(err) => {
+			eprintln!("{}", err);
+			return None;
+		},
+	}
+}
+
+fn create_compiled_contract_path(path_to_code: &PathBuf) -> PathBuf {
```

Comment:
What if multiple files in the same project have the same name, but are under different parent directories? They will overwrite one another cache file, effectively negating the effectiveness of caching.
We have to use a more precise way to store them.
I can see two methods:
- mimicking the project fs by creating all parent directories under "compiled-cairo-files"
- creating the file name based on its full `path /foo/bar/file.cairo` become `foo__bar__file.cairo`

The second one save io by creating less files, but it also introduce a risk for overlapping with files like `/foo__bar/file.cairo`. Still I think it is good enough and should not happen in 99.999% of case

Reply:
i updated this part!

kindly check on it when you have the time @tdelabro!